### PR TITLE
Spark 8426

### PR DESCRIPTION
**Pull request by mwws**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * The interface to determine executor blacklist and node blacklist.
+ */
+trait BlacklistStrategy {
+  val timeout: Long
+  def getExecutorBlacklist(failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String]
+  def getNodeBlacklist(failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String]
+
+  // Default implementation to remove failure executors from HashMap based on given time period.
+  def recoverFailureExecutors(failureExecutors: mutable.HashMap[String, FailureStatus]): Unit = {
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: expiredExecutorsInBlackList

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import scala.collection.mutable
+
+import org.apache.spark.SparkConf
+import org.apache.spark.util.SystemClock
+
+/**
+ * The interface to determine executor blacklist and node blacklist.
+ */
+trait BlacklistStrategy {
+  val timeout: Long
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: assigne default value

Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: expireTimeInMillisecond

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/**
+ * A type of blacklist strategy:
+ *
+ * An executor will be in blacklist, if it failed more than "executorFailedThreshold" times.
+ * A node will be in blacklist, if there are more than "nodeFailedThreshold" executors on it
+ * in executor blacklist
+ */
+class ThresholdStrategy (
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: SimpleStrategy?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    executorFailedThreshold: Int,
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: do you mean maxFailureTaskNumber?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala (hunk 2 above)
Comment (mwws) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: default value is provided in the apply method in "object BlackStrategy" which is come from SparkConf. Do we still need default value here? or maybe just add a comment to explain it?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    nodeFailedThreshold: Int,
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: maxBlackExecutorNumber

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    timeout: Long
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: expireTimeInMillisecond

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  )extends BlacklistStrategy {
+
+  private def executorBlacklistCandidate(
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: getSelectedExecutorMap

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      failureExecutors: mutable.HashMap[String, FailureStatus]) = {
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: failedExecutorMap

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+      .filter {case (host, failureExecutors) => failureExecutors.size > nodeFailedThreshold}
+      .keys.toSet
+  }
+}
+
+/**
+ * A type of blacklist strategy:
+ * Once task failed  at executor, put the executor and its node into blacklist.
+ */
+class StrictStrategy (timeout: Long) extends BlacklistStrategy {
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: NaiveStrategy?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+object BlacklistStrategy {
+  // Generate BlacklistStrategy based on spark configuration
+  def apply(sparkConf: SparkConf): BlacklistStrategy = {
+    val timeout = sparkConf.getLong("spark.scheduler.blacklist.timeout", 1800000L)
+    sparkConf.get("spark.scheduler.blacklist.strategy", "threshold") match {
+      case "threshold" =>
+        new ThresholdStrategy(
+            sparkConf.getInt("spark.scheduler.blacklist.threshold.executorFailedThreshold", 3),
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: mamAllowedFailureNumber?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+ */
+class BlacklistTracker(sparkConf: SparkConf) {
+  // maintain a ExecutorId --> FailureStatus HashMap
+  private val failureExecutors: mutable.HashMap[String, FailureStatus] = mutable.HashMap()
+
+  // Apply Strategy pattern here to change different blacklist detection logic
+  private val strategy = BlacklistStrategy(sparkConf)
+
+  // A daemon thread to expire blacklist executor periodically
+  private val executor = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala]: scheduler ?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 12]
```diff
@@ -259,7 +248,7 @@ private[spark] class TaskSetManager(
     while (indexOffset > 0) {
       indexOffset -= 1
       val index = list(indexOffset)
-      if (!executorIsBlacklisted(execId, index)) {
+      if (!blacklistTracker.fold(false)(_.executorIsBlacklisted(execId, sched))) {
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:251]: why we need fold here?  it seems tracker.isBlacklisted(execId, sched) == true/false is enough, right?

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  def stop(): Unit = {
+    executor.shutdown()
+    executor.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  def updateFailureExecutors(info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    reason match {
+      case Success => Unit
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala]: if Task success , to remove that executor from your failure list if exists

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 14]
```diff
@@ -641,7 +618,7 @@ private[spark] class TaskSetManager(
       logInfo("Ignoring task-finished event for " + info.id + " in stage " + taskSet.id +
         " because task " + index + " has already completed successfully")
     }
-    failedExecutors.remove(index)
+    blacklistTracker.foreach(_.removeFailureExecutors(Iterable(info.executorId)))
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: blacklistTracker.foreach(_.updateFailureExecutors(info, reason)) or to move this function to higher level

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 15]
```diff
@@ -434,7 +434,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size)
     }
 
-    doKillExecutors(executorsToKill)
+    val killRequestAcknowledged = doKillExecutors(executorsToKill)
+    if(killRequestAcknowledged) {
+      scheduler.sc.blacklistTracker.foreach(_.removeFailureExecutors(executorsToKill))
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: killed not means that the executor removed.

Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: to add that in removeExecutor

File: core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+      .filter {case (host, failureExecutors) => failureExecutors.size > nodeFailedThreshold}
+      .keys.toSet
+  }
+}
+
+/**
+ * A type of blacklist strategy:
+ * Once task failed  at executor, put the executor and its node into blacklist.
+ */
+class StrictStrategy (val timeout: Long) extends BlacklistStrategy {
```
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala]: it seems strickStrategy is only the example for simpleStrategy by setting number as 1. 

you can remove that or provide an alias for the example.

**Pull request merged**